translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,301 @@
+// limitations under the License.
+
+package metrics
+
+import (
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+	otlpresource "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := resourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func resourceToOC(resource *otlpresource.Resource) (*occommon.Node, *ocresource.Resource) {
+	// TODO extract from trace translator
+	panic("Not implemented")
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor),
+		Timeseries:       dataPointsToTimeseries(metric),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor) *ocmetrics.MetricDescriptor {
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        desctiptorTypeToOC(descriptor.Type),
+	}
+}
+
+func desctiptorTypeToOC(t otlpmetrics.MetricDescriptor_Type) ocmetrics.MetricDescriptor_Type {
+	switch t {
+	case otlpmetrics.MetricDescriptor_UNSPECIFIED:
+		return ocmetrics.MetricDescriptor_UNSPECIFIED
+	case otlpmetrics.MetricDescriptor_GAUGE_INT64:
+		return ocmetrics.MetricDescriptor_GAUGE_INT64
+	case otlpmetrics.MetricDescriptor_GAUGE_DOUBLE:
+		return ocmetrics.MetricDescriptor_GAUGE_DOUBLE
+	case otlpmetrics.MetricDescriptor_GAUGE_HISTOGRAM:
+		return ocmetrics.MetricDescriptor_GAUGE_DISTRIBUTION
+	case otlpmetrics.MetricDescriptor_COUNTER_INT64:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_INT64
+	case otlpmetrics.MetricDescriptor_COUNTER_DOUBLE:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_DOUBLE
+	case otlpmetrics.MetricDescriptor_CUMULATIVE_HISTOGRAM:
+		return ocmetrics.MetricDescriptor_CUMULATIVE_DISTRIBUTION
```

Comment:
@bogdandrutu @jmacd can you please verify that this correctly maps OpenTelemetry metric types to OpenCensus metric types?

Reply:
Yes, it is correct.